bench-vortex/src/bin/clickbench.rs

```diff
@@ -105,181 +113,414 @@ fn main() -> anyhow::Result<()> {
                 }
-            }
-            render_table(all_measurements, &args.formats, RatioMode::Time).unwrap()
-        }
-        DisplayFormat::GhJson => print_measurements_json(all_measurements).unwrap(),
+
+                fastest.min(duration)
+            })
+        });
+
+    (
+        fastest_run,
+        execution_plan
+            .into_inner()
+            .expect("Execution plan must be set"),
+    )
+}
+
+async fn execute_datafusion_query(
+    query_idx: usize,
+    query_string: &str,
+    iteration: usize,
+    session_context: datafusion::execution::context::SessionContext,
+) -> anyhow::Result<(Duration, std::sync::Arc<dyn execution_plan::ExecutionPlan>)> {
+    let time_instant = Instant::now();
+    let query_string = query_string.to_owned();
+
+    let execution_plan = tokio::task::spawn(async move {
+        let (_, execution_plan) = execute_query(&session_context, &query_string)
+            .instrument(info_span!("execute_query", query_idx, iteration))
+            .await
+            .unwrap_or_else(|e| panic!("executing query {query_idx}: {e}"));
+
+        execution_plan
+    })
+    .await?;
+
+    Ok((time_instant.elapsed(), execution_plan))
+}
+
+/// Executes a single ClickBench query using DuckDB.
+///
+/// # Returns
+///
+/// The duration of the fastest execution
+#[allow(clippy::let_and_return)]
+fn benchmark_duckdb_query(
+    query_idx: usize,
+    query_string: &str,
+    iterations: usize,
+    runtime: &tokio::runtime::Runtime,
+    file_format: Format,
+    url: &Url,
+    single_file: bool,
+) -> Duration {
+    let fastest_run = (0..iterations).fold(Duration::from_millis(u64::MAX), |fastest, _| {
+        runtime.block_on(async {
+            let duration = execute_duckdb_query(
+                query_idx,
+                query_string,
+                &std::path::PathBuf::from(url.as_str()),
+                file_format,
+                single_file,
+            )
+            .await
+            .unwrap_or_else(|err| panic!("query: {query_idx} failed with: {err}"));
+
+            fastest.min(duration)
+        })
+    });
+
+    fastest_run
+}
+
+async fn execute_duckdb_query(
+    query_idx: usize,
+    query_string: &str,
+    data_path: &std::path::Path,
+    file_format: Format,
+    single_file: bool,
+) -> anyhow::Result<Duration> {
+    let query_file = tempfile::tempdir()?
+        .path()
+        .join(format!("query_{query_idx}.sql"));
+
+    fs::write(&query_file, query_string)?;
+
+    let extension = match file_format {
+        Format::Parquet => "parquet",
+        Format::OnDiskVortex => "vortex",
+        other => vortex_panic!("Format {other} isn't supported on ClickBench"),
+    };
+
+    let file_glob = if single_file {
+        format!("{}/{extension}/hits.{extension}", data_path.display())
+    } else {
+        format!("{}/{extension}/*.{extension}", data_path.display())
+    };
+
+    let time_instant = Instant::now();
+
+    // TODO: complete duckdb setup
```

Comment:
shall we abort for now if engine is duckdb?

Reply:
Handled in `init_data_source` for now.